### PR TITLE
Introduce global PID

### PR DESCRIPTION
**Pull request by halilozanakgul**
DESCRIPTION: Introduces global pid

This PR introduces a process id for queries across the cluster called global pid. The global pid of a query and its distributed parts are the same. So, for a `SELECT` query, the coordinator query and the queries in the shards will have the same global pid, indicating they belong the same process.

A global pid is assigned to all Citus backends in `StartupCitusBackend` function.
Global pid is calculated (node id of the node the query started in) * 10<sup>10</sup> + (the pid of the query in this node).
A query with pid 15 that started in the node with id 2 would have the global pid 20000000015.

Global pid is added to the `get_all_active_transactions`, `get_global_active_transactions`, `citus_dist_stat_activity`, and `citus_worker_stat_activity` UDFs (and related views). Also, `get_all_active_transactions` and `get_global_active_transactions` now return all backends that have a global pid assigned.

### CODE REVIEW

File: src/backend/distributed/transaction/backend_data.c [hunk 1]
```diff
@@ -779,6 +781,31 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+		return;
+	}
+
+	StringInfo applicationName = makeStringInfo();
+	appendStringInfoString(applicationName, application_name);
+
+	char *globalPIDString = &applicationName->data[6];
+	uint64 globalPID = strtoul(globalPIDString, NULL, 10);
+
+	globalPID += 123;
```
Comment (marcocitus) [inline src/backend/distributed/transaction/backend_data.c:798]: what's 123?

Comment (halilozanakgul) [reply src/backend/distributed/transaction/backend_data.c]: This was just for debugging 😄, removed it

File: src/backend/distributed/transaction/backend_data.c [hunk 2]
```diff
@@ -779,6 +782,30 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+		return;
+	}
+
+	StringInfo applicationName = makeStringInfo();
+	appendStringInfoString(applicationName, application_name);
+
+	char *globalPIDString = &applicationName->data[6];
+	uint64 globalPID = strtoul(globalPIDString, NULL, 10);
+
+	MyBackendData->globalPID = globalPID;
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:800]: we should always guard accesses to `MyBackendData` via the `MyBackendData->mutex`. Please see other usages in the code

File: src/backend/distributed/transaction/backend_data.c [hunk 3]
```diff
@@ -780,6 +784,59 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+		MyBackendData->globalPID = globalPID;
+		SpinLockRelease(&MyBackendData->mutex);
+
+		return;
+	}
+
+	StringInfo applicationName = makeStringInfo();
+	appendStringInfoString(applicationName, application_name);
+
+	char *globalPIDString = &applicationName->data[6];
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:807]: Directly accessing to `[6]` seems unusual. Can we use something else, like `strstr` function or such? A very quick idea I can think of:

```C

if (strstr(application_name, CITUS_APP_NAME_PREFIX))
char *globalPIDString = &applicationName->data[strnlen(CITUS_APP_NAME_PREFIX)];
```

File: src/backend/distributed/connection/connection_configuration.c [hunk 4]
```diff
@@ -234,6 +235,9 @@ GetConnParams(ConnectionHashKey *key, char ***keywords, char ***values,
 	 */
 	char nodePortString[12] = "";
 
+	StringInfo applicationName = makeStringInfo();
+	appendStringInfo(applicationName, "citus-%ld", GetGlobalPID());
```
Comment (onderkalaci) [inline src/backend/distributed/connection/connection_configuration.c:239]: lets use `CITUS_APPLICATION_NAME_PREFIX`

File: src/backend/distributed/transaction/backend_data.c [hunk 5]
```diff
@@ -780,6 +784,59 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+}
+
+
+/*
+ * GenerateGlobalPID generates the global process id for the current backend
+ */
+static uint64
+GenerateGlobalPID(void)
+{
+	return (((uint64) GetLocalGroupId()) * 10000000000) + getpid();
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:836]: would be good to explain a little more with a comment. What is `10000000000` or why do we add pid?Can this ever overflow?

File: src/include/distributed/connection_management.h [hunk 6]
```diff
@@ -30,6 +30,7 @@
 
 /* application name used for internal connections in Citus */
 #define CITUS_APPLICATION_NAME "citus"
+#define CITUS_APPLICATION_NAME_PREFIX "citus-"
```
Comment (onderkalaci) [inline src/include/distributed/connection_management.h:33]: do we still need `CITUS_APPLICATION_NAME`?

File: src/backend/distributed/connection/connection_management.c [hunk 7]
```diff
@@ -1448,7 +1448,10 @@ ShouldShutdownConnection(MultiConnection *connection, const int cachedConnection
 bool
 IsCitusInitiatedRemoteBackend(void)
 {
-	return application_name && strcmp(application_name, CITUS_APPLICATION_NAME) == 0;
+	return application_name &&
+		   strncmp(application_name, CITUS_APPLICATION_NAME_PREFIX,
+				   strlen(CITUS_APPLICATION_NAME_PREFIX)) == 0 &&
```
Comment (onderkalaci) [inline src/backend/distributed/connection/connection_management.c:1453]: seems hard to read. Can you simplify a bit? Maybe assign `strlen(CITUS_APPLICATION_NAME_PREFIX)` to a variable etc.

Also, `strisdigit_s(application_name + strlen(CITUS_APPLICATION_NAME_PREFIX), 20)` what is this? Where do we get 20?

**Review by onderkalaci — COMMENTED** [review]
I think we are on the right direction, but the string manipulation should be done in a little more explicit manner. Now, it is hard to follow what is going on.

In general, a good PR description would be useful

File: src/backend/distributed/transaction/backend_data.c (hunk 3 above)
Comment (halilozanakgul) [reply src/backend/distributed/transaction/backend_data.c:807]: I'll add `CITUS_APP_NAME_PREFIX` but I already check if the `applicationName` starts with `citus-` in the `IsCitusInitiatedRemoteBackend`. Is that really necessary?

File: src/include/distributed/connection_management.h (hunk 6 above)
Comment (halilozanakgul) [reply src/include/distributed/connection_management.h:33]: It is still the fallback application name.
But I'm not sure when it falls back

File: src/backend/distributed/transaction/backend_data.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/transaction/backend_data.c:807]: Well, I think it is not _strictly_ necessary, but accessing `[6]` is:
(a) hard to understand what is going on / what we are doing
(b) could be hard to debug/understand if we one day decide to change  `CITUS_APP_NAME_PREFIX`
(c) has a dependency to `IsCitusInitiatedRemoteBackend` call, which may get refactored

So, it is mostly a readability and some robustness to refactors issue from my perspective

Comment (marcocitus) [reply src/backend/distributed/transaction/backend_data.c:807]: related, can we have an ExtractGlobalPID function to maintain the string formatting logic in a single place?

File: src/backend/distributed/connection/connection_configuration.c (hunk 4 above)
Comment (marcocitus) [reply src/backend/distributed/connection/connection_configuration.c:239]: have we considered other formats? (e.g. `"citus gpid=%ld"`)

File: src/include/distributed/connection_management.h (hunk 6 above)
Comment (onderkalaci) [reply src/include/distributed/connection_management.h:33]: > fallback_application_name
Specifies a fallback value for the application_name configuration parameter. This value will be used if no value has been given for application_name via a connection parameter or the PGAPPNAME environment variable. Specifying a fallback name is useful in generic utility programs that wish to set a default application name but allow it to be overridden by the user.

So, given that we are always going to have an application name, I think it'd be nice to drop `CITUS_APPLICATION_NAME` and all references to it for simplicity.

File: src/backend/distributed/connection/connection_configuration.c (hunk 4 above)
Comment (onderkalaci) [reply src/backend/distributed/connection/connection_configuration.c:239]: Note that `application_name` has 63 chars limit:
```C
		len = pg_mbcliplen(ident, len, NAMEDATALEN - 1);
```

If you set it  > 63, PG will truncate automatically
```SQL
set application_name TO '1234567891234567891234567891234567891234567891234567891234567891';
NOTICE:  identifier "1234567891234567891234567891234567891234567891234567891234567891" will be truncated to "123456789123456789123456789123456789123456789123456789123456789"
SET
```

max `uint64` is `18446744073709551615`, which is just 20 chars. So, Marco's suggestion sounds a lot nicer and seems to work fine.

**Review by onurctirtir — COMMENTED** [review]
I wonder should we skip copying "application_name" from the global params (citus.node_conninfo) in the first step.

In the second step, we will append our own `application_name` that contains the global pid but not sure if it could be problematic to define `application_name` more than once, or the which one would be taken into account by libpq.

In either case, seems to worth adding a test if possible, but I guess it would be slightly hard setting `citus.node_conninfo` requires a server restart.

https://github.com/citusdata/citus/blob/52879fdc96bf97aa5cdeea276eab2969f19cbef4/src/backend/distributed/connection/connection_configuration.c#L300-L318

**Comment by onurctirtir** [comment]
Should we first merge a smaller pr that doesn't append global pid to the `application_name` but still explicitly specifies `application_name` when preparing connection params so that we can backport into 10.1 & 10.2 ?

File: src/backend/distributed/transaction/backend_data.c [hunk 8]
```diff
@@ -780,6 +784,85 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+	return (((uint64) GetLocalNodeId()) * 10000000000) + getpid();
+}
+
+/*
+ * ExtractGlobalPID extracts the global process id from the application name  
+ */
+static uint64
+ExtractGlobalPID(void)
+{
+	StringInfo applicationName = makeStringInfo();
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:856]: Shouldn't we call `AssignGlobalPID` to make sure it is assigned?

File: src/backend/distributed/connection/connection_management.c [hunk 9]
```diff
@@ -1448,7 +1448,28 @@ ShouldShutdownConnection(MultiConnection *connection, const int cachedConnection
...
+	/* does application name exist */
+	if (!application_name)
+	{
+		return false;
+	}
+
+	uint64 prefixLength = strlen(CITUS_APPLICATION_NAME_PREFIX); 
+
+	/* does application name start with Citus's application name prefix */
+	if (strncmp(application_name, CITUS_APPLICATION_NAME_PREFIX, prefixLength) != 0)
```
Comment (onderkalaci) [inline src/backend/distributed/connection/connection_management.c:1460]: can we use `ExtractGlobalPID` here? I mean, change `ExtractGlobalPID` to get a `char *` and say return `NULL` if cannot be extracted.

And, if returns NULL we return false here?

File: src/test/regress/sql/metadata_sync_helpers.sql [hunk 10]
```diff
@@ -344,7 +344,7 @@ ROLLBACK;
 -- check with non-existing object type
 BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
 	SELECT assign_distributed_transaction_id(0, 8, '2021-07-09 15:41:55.542377+02');
-	SET application_name to 'citus';
+	SET application_name to 'citus gpid=10000000001';
```
Comment (onderkalaci) [inline src/test/regress/sql/metadata_sync_helpers.sql:347]: or maybe we can add a UDF to src/test folder such that `citus_assign_global_pid` which simply calls `AssignGlobalPID`

File: src/backend/distributed/transaction/backend_data.c [hunk 11]
```diff
@@ -780,6 +784,85 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
 
+/*
+ * AssignGlobalPID assigns a global process id for the current backend
+ */
+void
+AssignGlobalPID(void)
+{
+	if (MyBackendData->globalPID != 0)
+	{
+		elog(DEBUG1, "a global pid is already assigned to this backend");
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:795]: maybe print the globalId as well?And probably DEBUG4 is more convenient for such a low level information

File: src/backend/distributed/transaction/backend_data.c [hunk 12]
```diff
@@ -780,6 +784,85 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+}
+
+
+/*
+ * GetGlobalPID returns the global process id of the current backend
+ */
+uint64
+GetGlobalPID(void)
+{
+	uint64 globalPID = 0;
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:824]: what if `AssignGlobalPID` has not been called when this function is called? Should we maybe start the function with `AssignGlobalPID ` ?

File: src/backend/distributed/transaction/backend_data.c (hunk 8 above)
Comment (halilozanakgul) [reply src/backend/distributed/transaction/backend_data.c:856]: Ok I do this and remove the message in the assigning function that says "gpid is already assigned"

File: src/backend/distributed/connection/connection_management.c (hunk 9 above)
Comment (halilozanakgul) [reply src/backend/distributed/connection/connection_management.c:1460]: I'm not loving this idea. I would think it'd be the other way around.
Like calling this function (`IsCitusInitiatedRemoteBackend`) before (or in) `ExtractGlobalPID`.
If you insist on this can you explain your reasoning behind it?

File: src/backend/distributed/transaction/backend_data.c (hunk 11 above)
Comment (halilozanakgul) [reply src/backend/distributed/transaction/backend_data.c:795]: As I explained in the previous comment, I think we should remove this message since we decided to call it every time we call get gpid function

File: src/backend/distributed/transaction/backend_data.c (hunk 12 above)
Comment (halilozanakgul) [reply src/backend/distributed/transaction/backend_data.c:824]: I'll call the `AssignGlobalPID`.

File: src/backend/distributed/connection/connection_management.c (hunk 9 above)
Comment (onderkalaci) [reply src/backend/distributed/connection/connection_management.c:1460]: The idea is to follow Marco's earlier comment, have a single place where we know how to `ExtractGlobalPID` from a given string. Now, we have two, one in `ExtractGlobalPID` the other here. 

> Like calling this function (IsCitusInitiatedRemoteBackend) before (or in) ExtractGlobalPID.

This sounds counterintuitive. Why a utility function like `ExtractGlobalPID` should depend to a lot more generic function `IsCitusInitiatedRemoteBackend`?   

To be able to follow my suggestion, `ExtractGlobalPID` should be changed to a sole utility function, which takes a `char *` and returns globalPID (if not found returns 0/NULL).

File: src/backend/distributed/transaction/backend_data.c (hunk 8 above)
Comment (halilozanakgul) [reply src/backend/distributed/transaction/backend_data.c:862]: I realized that we shouldn't call `AssignGlobalPID`. Because this function doesn't actually look at the current backend's global pid. It extracts the gpid from the application name. So assigning won't do anything.

File: src/backend/distributed/connection/connection_management.c (hunk 9 above)
Comment (halilozanakgul) [reply src/backend/distributed/connection/connection_management.c]: We had offline discussion and decided to continue as @onderkalaci suggested

File: src/test/regress/sql/metadata_sync_helpers.sql (hunk 10 above)
Comment (halilozanakgul) [reply src/test/regress/sql/metadata_sync_helpers.sql:347]: That would assign the gpid to the backend but not set the `application_name`. I don't think it'd be enough.

File: src/include/distributed/connection_management.h [hunk 13]
```diff
@@ -28,8 +28,8 @@
 /* used for libpq commands that get an error buffer. Postgres docs recommend 256. */
 #define ERROR_BUFFER_SIZE 256
 
-/* application name used for internal connections in Citus */
-#define CITUS_APPLICATION_NAME "citus"
+/* application name prefix used for internal connections in Citus */
+#define CITUS_APPLICATION_NAME_PREFIX "citus gpid="
```
Comment (marcocitus) [inline src/include/distributed/connection_management.h:32]: can there be cases where we connect as "citus", but without a gpid (e.g. deadlock detector?)

should we just use "citus" in that case? (or a custom name?)

File: src/backend/distributed/sql/udfs/citus_dist_stat_activity/latest.sql [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+                                                               OUT pid int, OUT usesysid oid, OUT usename name, OUT application_name text, OUT client_addr INET,
+                                                               OUT client_hostname TEXT, OUT client_port int, OUT backend_start timestamptz, OUT xact_start timestamptz,
+                                                               OUT query_start timestamptz, OUT state_change timestamptz, OUT wait_event_type text, OUT wait_event text,
+                                                               OUT state text, OUT backend_xid xid, OUT backend_xmin xid, OUT query text, OUT backend_type text, OUT global_pid int8)
+RETURNS SETOF RECORD
+LANGUAGE C STRICT AS 'MODULE_PATHNAME',
+$$citus_dist_stat_activity$$;
+
+COMMENT ON FUNCTION pg_catalog.citus_dist_stat_activity(OUT query_hostname text, OUT query_hostport int, OUT distributed_query_host_name text, OUT distributed_query_host_port int,
+                                                        OUT transaction_number int8, OUT transaction_stamp timestamptz, OUT datid oid, OUT datname name,
```
Comment (marcocitus) [inline src/backend/distributed/sql/udfs/citus_dist_stat_activity/latest.sql:14]: this can probably be a separate PR, but even I have no idea what many of these fields mean and what the user would do would them.

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql [hunk 15]
```diff
@@ -13,6 +13,56 @@
...
 #include "udfs/worker_drop_sequence_dependency/11.0-1.sql"
 
+#include "udfs/get_all_active_transactions/11.0-1.sql"
+#include "udfs/get_global_active_transactions/11.0-1.sql"
+#include "udfs/citus_worker_stat_activity/11.0-1.sql"
+#include "udfs/citus_dist_stat_activity/11.0-1.sql"
+
+SET search_path = 'pg_catalog';
+
+CREATE VIEW citus.citus_lock_waits AS
```
Comment (onderkalaci) [inline src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:23]: why are we re-creating the view? Have we changed anything on the view's definition?

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql [hunk 16]
```diff
@@ -13,6 +13,56 @@
...
+  SELECT
+  citus_dist_stat_activity.*, (CASE citus_dist_stat_activity.distributed_query_host_name WHEN 'coordinator_host' THEN 0 ELSE pg_dist_node.nodeid END) as initiator_node_id
+  FROM
+  citus_dist_stat_activity LEFT JOIN pg_dist_node
+  ON
+  citus_dist_stat_activity.distributed_query_host_name = pg_dist_node.nodename AND
+  citus_dist_stat_activity.distributed_query_host_port = pg_dist_node.nodeport
+)
+SELECT
+ waiting.pid AS waiting_pid,
```
Comment (onderkalaci) [inline src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:45]: I think we should consider showing the `globalPID` of both processes as well.

File: src/backend/distributed/sql/udfs/citus_dist_stat_activity/latest.sql (hunk 14 above)
Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_dist_stat_activity/latest.sql:14]: Also, we could move `OUT global_pid int8` to earlier on the output. But, I also think that a separate PR would be more reasonable as these changes are too verbose and could make the actual code review more complicated :)

File: src/test/regress/expected/isolation_get_all_active_transactions.out [hunk 17]
```diff
@@ -35,8 +35,8 @@ step s2-begin-insert:
 
 step s3-as-admin:
  -- Admin should be able to see all transactions
- SELECT count(*) FROM get_all_active_transactions();
- SELECT count(*) FROM get_global_active_transactions();
+ SELECT count(*) FROM get_all_active_transactions() WHERE datid != 0;
```
Comment (onderkalaci) [inline src/test/regress/expected/isolation_get_all_active_transactions.out:38]: Why we needed these changes, do we get datid == 0 rows back?

File: src/test/regress/expected/isolation_global_pid.out [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+step s1-select:
+    SELECT * FROM dist_table;
+
+a|b
+---------------------------------------------------------------------
+(0 rows)
+
+step s2-citus_worker_stat_activity:
+    SELECT count(*) FROM citus_worker_stat_activity() WHERE global_pid IN (
```
Comment (onderkalaci) [inline src/test/regress/expected/isolation_global_pid.out:38]: hmm, we set force_max_query_parallelization TO ON, shouldn't we get #shards worker queries? Is the shard count == 2?

File: src/backend/distributed/transaction/backend_data.c [hunk 19]
```diff
@@ -780,6 +805,113 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+	return (((uint64) GetLocalNodeId()) * 10000000000) + getpid();
+}
+
+
+/*
+ * ExtractGlobalPID extracts the global process id from the application name and returns it
+ * if the application name is not compatible with Citus' application names returns 0
+ */
+uint64
+ExtractGlobalPID(void)
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:882]: I think `ExtractGlobalPID` should get `application_name` as a parameter. Otherwise, it looks vague that we *extract* anything for the readers & other parts of the code.

File: src/backend/distributed/transaction/backend_data.c [hunk 20]
```diff
@@ -780,6 +805,113 @@ MarkCitusInitiatedCoordinatorBackend(void)
 }
 
 
+/*
+ * AssignGlobalPID assigns a global process id for the current backend if
+ * it is not already assigned
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:810]: might use some comment. When do we generate new one? When do we use the globalPID assigned to this process?

File: src/backend/distributed/transaction/backend_data.c [hunk 21]
```diff
@@ -780,6 +805,113 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+		uint64 globalPID = GenerateGlobalPID();
+
+		SpinLockAcquire(&MyBackendData->mutex);
+		MyBackendData->globalPID = globalPID;
+		SpinLockRelease(&MyBackendData->mutex);
+
+		return;
+	}
+
+	uint64 globalPID = ExtractGlobalPID();
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:831]: Minor:
For example, `ExtractGlobalPID(application_name)` would be much nicer to follow the code.

Also,  adding an `else` seems to increase readability `if( IsCitusInitiatedRemoteBackend) (Extrat) else {Generate}`

File: src/backend/distributed/transaction/backend_data.c [hunk 22]
```diff
@@ -780,6 +805,113 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+
+
+/*
+ * GetGlobalPID returns the global process id of the current backend
+ */
+uint64
+GetGlobalPID(void)
+{
+	/* assign here incase it is not assigned already */
+	AssignGlobalPID();
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:846]: ok, as we call `AssignGlobalPID ` in `StartupCitusBackend`, calling it in other places seems redundant / confusing.


`StartupCitusBackend` guaranteed to be called for all Citus backends, hence we should not try to re-assign.

Similarly, the early return in `AssignGlobalPID` with `MyBackendData->globalPID != 0` seems redundant and error prone as well.

We should at most convert them to asserts.

(I know that was my idea, sorry for the extra iteration)

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql (hunk 15 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:24]: It's because we drop a UDF this view depends. I'll put a comment explaining this.

Comment (halilozanakgul) [reply src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:24]: Ok but when you drop the function this view depends it drops the view too.

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql [hunk 23]
```diff
@@ -14,6 +14,56 @@
 #include "udfs/worker_drop_sequence_dependency/11.0-1.sql"
 #include "udfs/worker_drop_shell_table/11.0-1.sql"
 
+#include "udfs/get_all_active_transactions/11.0-1.sql"
```
Comment (onderkalaci) [inline src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:17]: #include "udfs/get_all_active_transactions/11.0-1.sql"
create view

....

File: src/test/regress/expected/isolation_global_pid.out [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+step s1-worker-begin:
+    SELECT run_commands_on_session_level_connection_to_node('BEGIN');
+
+run_commands_on_session_level_connection_to_node
+---------------------------------------------------------------------
+
+(1 row)
+
+step s1-worker-select:
+    SELECT run_commands_on_session_level_connection_to_node('SET citus.force_max_query_parallelization TO ON; SELECT * FROM dist_table');
```
Comment (onderkalaci) [inline src/test/regress/expected/isolation_global_pid.out:26]: SET citus.enable_local_execution TO false until we fix: https://github.com/citusdata/citus/issues/5282

File: src/backend/distributed/transaction/backend_data.c [hunk 25]
```diff
@@ -780,6 +805,108 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+
+	/* does application name start with Citus's application name prefix */
+	if (strncmp(applicationNameCopy->data, CITUS_APPLICATION_NAME_PREFIX, prefixLength) != 0)
+	{
+		return 0;
+	}
+
+	/* are the remaining characters of the application name numbers */
+	uint64 numberOfRemainingChars = strlen(applicationNameCopy->data) - prefixLength;
+	if (numberOfRemainingChars == 0 ||
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:897]: let be more defensive: ```numberOfRemainingChars <= 0```

File: src/backend/distributed/transaction/backend_data.c [hunk 26]
```diff
@@ -780,6 +805,108 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+uint64
+ExtractGlobalPID(char* applicationName)
+{
+	/* does application name exist */
+	if (!applicationName)
+	{
+		return 0;
+	}
+
+	// we create our own copy of application name incase the original changes
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:883]: minor: the comment style should be : /**/

File: src/include/distributed/connection_management.h (hunk 13 above)
Comment (onderkalaci) [inline src/include/distributed/connection_management.h:32]: needs rebase as we have changed it to `citus_internal`. I think, I'd stick to `citus_internal`

File: src/test/regress/expected/failure_connection_establishment.out [hunk 27]
```diff
@@ -225,8 +225,8 @@ SELECT count(*) FROM single_replicatated WHERE key = 100;
 
 RESET client_min_messages;
 -- verify get_global_active_transactions works when a timeout happens on a connection
-SELECT get_global_active_transactions();
- get_global_active_transactions
+SELECT * FROM get_global_active_transactions() WHERE datid != 0;
```
Comment (onderkalaci) [inline src/test/regress/expected/failure_connection_establishment.out:228]: as we discussed, lets use another field than `datid`

Comment (onderkalaci) [reply src/test/regress/expected/failure_connection_establishment.out:228]: And, for reference, we have to add these because now maintenance daemon processes shows up in `get_global_active_transactions/get_all_active_transactions`. 

We don't mind `get_global_active_transactions/get_all_active_transactions`.  But, they also current show up in citus_dist/worker_stat_activity views, where we'd like to avoid maintenance processes. That is going to happen with another PR

File: src/test/regress/expected/metadata_sync_helpers.out [hunk 28]
```diff
@@ -1277,7 +1277,7 @@ BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
 
 (1 row)
 
-	SET application_name to 'citus';
+	SET application_name to 'citus gpid=10000000001';
```
Comment (onderkalaci) [inline src/test/regress/expected/metadata_sync_helpers.out:1280]: maybe add one tests where `citus gpid=test` which throws error?

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql (hunk 15 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:25]: Added comment explaining this

File: src/test/regress/expected/isolation_get_all_active_transactions.out (hunk 17 above)
Comment (halilozanakgul) [reply src/test/regress/expected/isolation_get_all_active_transactions.out]: I'm updating to transaction_number

File: src/include/distributed/connection_management.h (hunk 13 above)
Comment (halilozanakgul) [reply src/include/distributed/connection_management.h:32]: I'm making it "citus_internal gpid=" then

**Review by onderkalaci — COMMENTED** [review]
Can you also please add some explanation on what we achieve with the global PIDs to the PR description?

Overall, it is almost ready to merge, after considering the small notes and doing some further testing

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql (hunk 23 above)
Comment (marcocitus) [reply src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:17]: Would be nicer to also put citus_lock_waits in a separate file to track changes.

File: src/backend/distributed/metadata/metadata_cache.c [hunk 29]
```diff
@@ -3618,6 +3621,60 @@ GetLocalGroupId(void)
...
+	 * localGroupId in pg_dist_node
+	 */
+	if (nodeId == -1)
+	{
+		elog(DEBUG1, "there is no active node with group id '%d' on pg_dist_node",
+			 localGroupId);
+		if (IsCoordinator())
+		{
+			nodeId = 0;
+		}
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/metadata/metadata_cache.c:3669]: I will do that, but if current node is a worker node and is not in the `pg_dist_node` I think there are bigger problems.

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql (hunk 23 above)
Comment (onderkalaci) [reply src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:17]: Yeah, we considered that, but files are under `udfs` folder and this is a view. So, we weren't sure if it'd make sense to put there.

Ozan will quickly check if we can create `views` folder. If not, we may discuss whether we want to work on adding `views` folder or keep it here.

File: src/backend/distributed/transaction/backend_data.c [hunk 30]
```diff
@@ -780,6 +806,109 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+	if (strncmp(applicationNameCopy->data,
+			    CITUS_APPLICATION_NAME_PREFIX, prefixLength) != 0)
+	{
+		return INVALID_CITUS_INITIATED_BACKEND_PID;
+	}
+
+	/* are the remaining characters of the application name numbers */
+	uint64 numberOfRemainingChars = strlen(applicationNameCopy->data) - prefixLength;
+	if (numberOfRemainingChars <= 0 ||
+		!strisdigit_s(applicationNameCopy->data + prefixLength, numberOfRemainingChars))
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/transaction/backend_data.c:899]: After offline discussion we decided to put these tests to `metadata_sync_helpers.sql`

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql (hunk 23 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:17]: Our current convention, although not ideal, is to create views in the `udfs` file. So I did the same.
And created an issue to fix this: #5691

Comment (onderkalaci) [reply src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:17]: Thanks for acting on this

File: src/backend/distributed/transaction/backend_data.c [hunk 31]
```diff
@@ -384,8 +390,7 @@ StoreAllActiveTransactions(Tuplestorestate *tupleStore, TupleDesc tupleDescripto
 
 		SpinLockAcquire(&currentBackend->mutex);
 
-		/* we're only interested in backends initiated by Citus */
-		if (currentBackend->citusBackend.initiatorNodeIdentifier < 0)
+		if (currentBackend->globalPID == 0)
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:393]: maybe use INVALID_CITUS_INTERNAL_BACKEND_GPID

File: src/test/regress/spec/isolation_global_pid.spec [hunk 32]
```diff
@@ -0,0 +1,95 @@
...
+step "s2-coordinator-get_global_active_transactions"
+{
+    SELECT count(*) FROM get_global_active_transactions() WHERE global_pid IN (
+        SELECT global_pid FROM citus_dist_stat_activity() WHERE query LIKE '%SELECT * FROM dist\_table%'
+    );
+}
+
+
+// worker - coordinator
+permutation "s1-start-session-level-connection" "s1-worker-begin" "s1-worker-select" "s2-coordinator-citus_dist_stat_activity" "s2-coordinator-citus_worker_stat_activity" "s1-worker-commit" "s1-stop-session-level-connection"
```
Comment (onderkalaci) [inline src/test/regress/spec/isolation_global_pid.spec:92]: why don't we have worker - worker? I think it'd be useful to have worker-worker excluding the active transactions UDFs.

**Review by onderkalaci — APPROVED** [review]
After two minor suggestions, we are good to 🚢

File: src/backend/distributed/transaction/backend_data.c [hunk 33]
```diff
@@ -780,6 +805,109 @@ MarkCitusInitiatedCoordinatorBackend(void)
...
+ExtractGlobalPID(char *applicationName)
+{
+	/* does application name exist */
+	if (!applicationName)
+	{
+		return INVALID_CITUS_INTERNAL_BACKEND_GPID;
+	}
+
+	/* we create our own copy of application name incase the original changes */
+	StringInfo applicationNameCopy = makeStringInfo();
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/backend_data.c:884]: One minor comment: `makeStringInfo` allocates 1024 bytes, https://github.com/postgres/postgres/blob/27b77ecf9f4d5be211900eda54d8155ada50d696/src/common/stringinfo.c#L39-L64

So, maybe
```C
char *applicationNameCopy = pstrdup(applicationName);


....

```

Such that we allocate at most 64 chars, which is the most `applicationName` can be

File: src/test/regress/spec/isolation_global_pid.spec (hunk 32 above)
Comment (halilozanakgul) [reply src/test/regress/spec/isolation_global_pid.spec:92]: Our isolation testing structure doesn't show results of `SELECT` queries on workers.
